src/TestResults/match-by-context.ts

```diff
@@ -0,0 +1,292 @@
+};
+export const toMatchResult = (
+  test: ItBlock,
+  assertionOrErr: TestAssertionStatus | string
+): TestResult => {
+  const assertion = typeof assertionOrErr === 'string' ? undefined : assertionOrErr;
+  const err = typeof assertionOrErr === 'string' ? assertionOrErr : undefined;
+
+  // Note the shift from one-based to zero-based line number and columns
+  return {
+    name: test.name,
+    start: {
+      column: test.start.column - 1,
+      line: test.start.line - 1,
+    },
+    end: {
+      column: test.end.column - 1,
+      line: test.end.line - 1,
+    },
+
+    status: assertion?.status ?? TestReconciliationState.Unknown,
+    shortMessage: assertion?.shortMessage ?? err,
+    terseMessage: assertion?.terseMessage,
+    lineNumberOfError: matchPos(test, assertion, false) ? assertion.line - 1 : test.end.line - 1,
+  };
+};
+
+/** mark all data and child containers unmatched */
+const toUnmatchedResults = (tContainer: ContainerNode<ItBlock>, err: string): TestResult[] => {
+  const results = tContainer.childData.map((n) => toMatchResult(n.only(), err));
+  tContainer.childContainers.forEach((c) => results.push(...toUnmatchedResults(c, err)));
+  return results;
+};
+
+type MessageType = 'context-mismatch' | 'match-failed' | 'unusual-match' | 'duplicate-test-name';
+const createMessaging = (fileName: string, verbose: boolean) => (
+  messageType: MessageType,
+  contextType: ContextType,
+  source: ContainerNode<ItBlock> | DataNode<ItBlock>,
+  assertion?: ContainerNode<TestAssertionStatus> | DataNode<TestAssertionStatus>,
+  extraReason?: string
+): void => {
+  if (!verbose) {
+    return;
+  }
+
+  const output = (message: string): void =>
+    console.warn(`[${fileName}] ${message} \n source=`, source, `\n assertion=`, assertion);
+
+  const blockType = contextType === 'container' ? 'describe' : 'test';
+  switch (messageType) {
+    case 'context-mismatch':
+      output(
+        `!! context mismatched !! ${contextType} nodes are different under "${source.name}": `
+      );
+      break;
+    case 'match-failed': {
+      output(`!! match failed !! ${blockType}: "${source.name}": ${extraReason} `);
+      break;
+    }
+    case 'duplicate-test-name': {
+      output(
+        `duplicate test names in the same describe block is not recommanded and might not be matched reliably: ${blockType} ${source.name}`
+      );
+      break;
+    }
+    case 'unusual-match': {
+      output(`unusual match: ${extraReason} : ${blockType} ${source.name}: `);
+      break;
+    }
+  }
+};
+type Messaging = ReturnType<typeof createMessaging>;
+
+/**
+ * create a handler to process matched test block: (test, assertion) pair
+ * @param config
+ * @return a function to match each it block
+ */
+const HandleTestBlockMatch = (config: {
+  messaging: Messaging;
+  onResult: (result: TestResult) => void;
+}) => (t: DataNode<ItBlock>, matched: DataNode<TestAssertionStatus>[]): void => {
+  const { messaging, onResult } = config;
```

Comment:
Is there any reason for not destructuring it already as part of the function argument?
```ts
const HandleTestBlockMatch = ({ messaging, onResult }: {
  messaging: Messaging;
  onResult: (result: TestResult) => void;
}) => (t: DataNode<ItBlock>, matched: DataNode<TestAssertionStatus>[]): void => {
```

Reply:
good point, but this morning I look at the code again and really didn't like the redundant setup either, so I tried a hybrid flattering approach, the goal is to reduce the "nesting" while still group related methods together. 

Also, I renamed the files to be all lower case as this is a pretty consistent suggestion on js file naming convention mainly to safeguard env that do not distinguish upper/lower cases.